### ISSUE TITLE
Aerospace WM support

### ISSUE THREAD

**Issue by AnirudhG07** [Jean-Tinland/simple-bar#423]
**Is your feature request related to a problem? Please describe.**
I really like the looks of Simple-bar, but as the repo says, its for yabai. So Does it work for Aerospace WM for Mac???

**Describe the solution you'd like**
If it doesn't work for aerospace, is there anyway I can use simple-bar?

**Describe alternatives you've considered**
NA

**Additional context**
I searched for any issues related to it, coudnt find.

**Comment by Jean-Tinland** [comment]
Hi, sorry for my late response!

Unfortunately, simple-bar is working only with yabai, as it depends on its signals and command line interface for spaces and application display.

I'm not planning to adapt simple-bar to Aeorspace for now but if this request get sufficient traction, I could take a deeper look on the feasibility :)

**Comment by AnirudhG07** [comment]
Thanks! It would be much appreciated to get Aerospace with simple-bar! And I think people will also like it.

**Comment by Zik42** [comment]
+1 to this feature

**Comment by Jean-Tinland** [comment]
From now on, AeroSpace is supported! It lacks some features compared to the yabai implementation but it is completely usable.

I will keep this issue opened in pin it so anyone can leave feedback and feature requests!

**Comment by AnirudhG07** [comment]
Lessgooo!

**Comment by AnirudhG07** [comment]
hi @Jean-Tinland, if I press `alt(option) + 1` to shift to `1` workspace, the workspace number of simple-bar doesnt change. 

![31521](https://github.com/user-attachments/assets/882fd7cf-7df4-479b-9427-1dc8b7ae777a)

How to fix this?
Also, I m really loving simple-bar! thanks so much!

**Comment by Jean-Tinland** [comment]
Glad to hear everything works fine now and happy to hear you like simple-bar! :)

**Comment by AnirudhG07** [comment]
I have a few questions I tried to looked into the docs(might have skipped)
1) Can the delay time of highlight in the bar when switching from workspace X to Y be reduced? when I switch, I see this 1/2-1 second delay bw the bar to load with new workspace info.
2) Can we move around the widget locations? Pls refer to the docs link if its present(sorry if I missed it)

**Comment by Jean-Tinland** [comment]
There is indeed a slight delay when switching spaces. 

It is mainly due to how Übersicht works: it can only be refreshed with the help of applescript (`osascript -e ...`). This process takes less or more time depending on your hardware. It is clearly not optimal for a widget like simple-bar that requires quick and repeated refreshes.

In order to circumvent that, I created [simple-bar-server](https://github.com/Jean-Tinland/simple-bar-server). This little node server is allowing to bypass Übersicht and communicate directly with all simple-bar's widgets with web sockets. It is a lot faster for the yabai implementation but it necessitate to install another tool so not everyone will be on board with this approach.

I'm planning to add the AeroSpace implementation in simple-bar-server soon! I'll let you know when everything is ready.

Using this basically allow you to refresh any simple-bar widget with simple `curl`.

About you second question, you can enable the "Bottom bar" setting in the "Global" tab.
If you need to change the data (right) widgets order, you can do it with the "Custom styles" tab and assign a new CSS `order` to each data widget.

For example:

```css
.weather {
  order: 2;
}

.battery {
  order: 1;
}
```

Let me know if this helps you.

**Comment by nikitabobko** [comment]
I don't know about simple-bar-server and Übersicht specific problems, but I'd like to notice that AeroSpace itself can be the cause of the delay.

The problem here is that AeroSpace is a single threaded application, and macOS AX API is blocking API, so when you switch workspaces, AeroSpace thread is blocked by macOS, and the thread doesn't have resources to answer the CLI query like `aerospace list-workspaces`.

I plan to address such performance problems in https://github.com/nikitabobko/AeroSpace/issues/104 https://github.com/nikitabobko/AeroSpace/issues/131

Currently, the workaround is to use `AEROSPACE_FOCUSED_WORKSPACE` environment variable (as we do it in integration with sketchybar https://nikitabobko.github.io/AeroSpace/goodies#show-aerospace-workspaces-in-sketchybar) and avoid 1 unnecessary `aerospace list-workspaces` query request.

I see that https://github.com/Jean-Tinland/simple-bar/commit/5852a58f7a8db90cbf335cc1a1d97b750d8ad59a doesn't use `AEROSPACE_FOCUSED_WORKSPACE`, so the delay may not be completely eliminated, even after "simple-bar-server fix"

**Comment by Jean-Tinland** [comment]
@nikitabobko I see! Thank you for all theses details! 
Unfortunately, it is not possible to pass arguments to Übersicht when asking it to refresh a specific widget.

I won't be able to improve the default implementation but when I'll work on simple-bar-server integration of AeroSpace, the use of the `AEROSPACE_FOCUSED_WORKSPACE` environment variable will be possible. Thank you again for this :)

**Comment by AnirudhG07** [comment]
Hi, is there a way to only show those workspaces on top which have some soft open? else it just disappears. Since aerospace has like multiple empty workspaces as well, and all of them just shows up at the top, even if empty, doesnt look neat.

**Comment by Jean-Tinland** [comment]
Sure, there is a "Hide empty spaces" setting in the "Spaces" tab in the settings.

**Comment by AnirudhG07** [comment]
Great! I think just like you have yabai based sections in docs, maybe you should have 1 for aerospace addressing these issues + Some nice Tips which aerospace users might benefit from! If you would like

**Comment by AnirudhG07** [comment]
Ques. I have toggled empty spaces to not be visible. But in aerospace I have many such spaces. This also disables if I am in an empty workspace. Say I am in `1`, but thats empty. then simple-bar doesnt show that I am in `1`. 
Is there anyway to enable it?

**Comment by Jean-Tinland** [comment]
That was a bug! Thank you for reporting it. It should be fixed in my latest commit.

**Comment by AnirudhG07** [comment]
Thanks! It works perfectly

**Comment by hld0** [comment]
Hi guys! First of all big thank you to Jean, I have been using simple-bar for so long that I don't even know how to go back to the old menu bar.

I recently made the switch from yabai to aerospace too, and played around with the config a bit. Today realized something that I don't know how to classify - I don't know if it is a bug or just an oversight of me somewhere in the config. But maybe it will help somebody facing this problem too:

My usual working setup includes two monitors (one Dell, one LG) with my MacBook lid being closed. Therefore, I used the following [assignment](https://nikitabobko.github.io/AeroSpace/guide#assign-workspaces-to-monitors) in the `.aerospace.toml` (I am not using spaces 1 and 2 as it clashes with other shortcuts otherwise)

```toml
 [workspace-to-monitor-force-assignment]
    3 = ["dell", "main"]
    4 = ["dell", "main"]
    5 = ["dell", "main"]
    6 = ["LG", "main"]
    7 = ["LG", "main"]
    8 = ["LG", "main"]
    9 = ["LG", "main"]
```

Now what happened was, that `simple-bar` showed the spaces 3,4,5 on the LG monitor and the spaces 6,7,8,9 on the Dell monitor. However, clicking on a space in the bar correctly switched the focus to the other monitor (e.g. I click on space 6 which is shown in the Dell monitor, the cursor correctly jumps to the LG monitor). 

What fixed this for me, was moving the menu bar in the display arrangement settings (`Settings -> Displays -> Arrange...`) from one screen to the other. You can click and drag the gray menu bar shown in one of the displays around:

![image](https://github.com/user-attachments/assets/8d243758-69fe-44c4-b3dc-36f05afa80a7)

Now the spaces are shown correctly in the simple bar too. Is this expected behavior?

**Comment by Jean-Tinland** [comment]
Hi @hld0! Thank you for your kind words :)

Thanks for explaining your configuration, I think this is due to a monitor number mismatch between simple-bar and AeroSpace. Because of this, simple-bar is displaying the workspaces on the wrong monitor.

I don't know the rules for the monitor/id association thus I can't explain what is going on. I had indeed some problems while integrating this so I settled on using `monitor-appkit-nsscreen-screens-id` for AeroSpace workspaces numbering.

_AeroSpace documentation excerpt:_

> %{monitor-id}
> 1-based Number. Sequential number of the belonging monitor

> %{monitor-appkit-nsscreen-screens-id}
> 1-based index of the belonging monitor in NSScreen.screens array. Useful for integration with other tools that might be using NSScreen.screens ordering (like sketchybar).

In my tests, I saw that these values were not the same depending mainly on the displays arrangement in Macos settings.

I could add a setting in simple-bar allowing to choose which value to use (`monitor-id` or `monitor-appkit-nsscreen-screens-id`). Maybe it would solve your issue but it would require you to test it.

As a last resort, your solution seems to work well even if it was not intended on my side.

Please let me know if you want me to implement this. If so, I'll also add an entry in documentation explaining the issue.

**Comment by haxybaxy** [comment]
Hello,

<img width="1470" alt="Image" src="https://github.com/user-attachments/assets/c14981e5-f1ca-425f-8933-509567171613" />

Sorry for bothering but I am having trouble with getting simplebar to start with aerospace.
I get the error:
```
simple-bar-index.jsx: aerospace is not running
```
Nothing is logged to the debug console, I used to get a 

```
[Error] XMLHttpRequest cannot load http://127.0.0.1:41416/run/ due to access control checks.
``` 
but It went away once I turned off the browser track widget

I am sure that aerospace is in fact running, restarting it, and doing ```aerospace enable off``` and ```aerospace enable on``` does not solve the issue.

I tried installing ubersicht from the website and brew, no change in the behaviour.

when I run ```which aerospace``` I get this path:
```
/opt/homebrew/bin/aerospace
```
which is the default path indicated in the settings module.

I also have the code snippet added to my aerospace.toml
```
on-focus-changed = [
  "exec-and-forget osascript -e 'tell application id \"tracesOf.Uebersicht\" to refresh widget id \"simple-bar-index-jsx\"'",
]

exec-on-workspace-change = [
  '/bin/zsh',
  '-c',
  '/usr/bin/osascript -e "tell application id \"tracesOf.Uebersicht\" to refresh widget id \"simple-bar-index-jsx\""',
]
```
I am on ubersicht 1.6 and MacOS Sequoia 15.1.1
Is there something I am missing? I would love to get this working
Thank you

**Comment by Jean-Tinland** [comment]
Hi! Can you check if you are using the latest version of AeroSpace?

**Comment by haxybaxy** [comment]
I updated aerospace from 15.2 to 16.2 and it fixed it, thank you.

**Comment by Jean-Tinland** [comment]
@FormalSnake These instructions are already in place [here](https://www.jeantinland.com/toolbox/simple-bar/documentation/installation/#or-your-aerospace-config).

Maybe they are not highlighted enough! Where did you look for them? I can rearrange the document if necessary. :)

**Comment by FormalSnake** [comment]
> @FormalSnake These instructions are already in place [here](https://www.jeantinland.com/toolbox/simple-bar/documentation/installation/#or-your-aerospace-config).
> 
> Maybe they are not highlighted enough! Where did you look for them? I can rearrange the document if necessary. :)

Yes i looked on the documentation, didn't think about checking the installation part because I got it running easily.
Might be nice to have it [here](https://www.jeantinland.com/toolbox/simple-bar/documentation/spaces-settings/) or directly have a common questions page with some specific instructions.
Thank you for the quick and friendly reply :D

**Comment by Jean-Tinland** [comment]
@FormalSnake I created a specific installation guide for each window manager. I think it will allow for a better user experience in this documentation.

I'll close this issue as AeroSpace implementation seems to work correctly. Feel free to open a new issue if you encounter any problem or create a discussion if you have any question.
